### PR TITLE
Allow configuring required address comonents

### LINKED ISSUE #47: Do not use `user_country` as the only address parameter for XCM
The `user_country` parameter gets submitted by Twingle with every call on TwingleDonation.submit. Because the `user_country` is passed to the [XCM](https://github.com/systopia/de.systopia.xcm), it gets interpreted as an address information what can lead to an [overwriting of existing address information](https://github.com/systopia/de.systopia.xcm/issues/78).

[One way to prevent this](https://github.com/systopia/de.systopia.twingle/pull/46) is to not check the "Change Primary Detail?" option in the XCM profile. But maybe one wants to update primary address information by the newer information from the Twingle call. With this solution,  this would not be possible any more. 

It would be preferable to do not use the `user_country` as address parameter if it's the only address information submitted.

This enhancement can be done with a few lines of code, I think.

### PR DESCRIPTION
**Pull request by MarcMichalsky**
This is a real KISS solution. And to be honest, I am not exceedingly proud of it. But it works and I think its simple nature makes it easily understandable for everyone.

I understand this PR as a voluntary act of mine to improve this cool open source software. It's just an offer and I do not demand you to pull it. So, there's no need to send a bill. 😬

fix #47

### CODE REVIEW

**Comment by MarcMichalsky** [comment]
What's that part for exactly?
https://github.com/systopia/de.systopia.twingle/blob/4659b535228856fdc9098fc6def456ecce8daf71/api/v3/TwingleDonation/Submit.php#L332-L334

**Comment by jensschuppe** [comment]
This removes e.g. the `user_street` element from the `$params` array, since its value is being set to the internal CiviCRM parameter name (`street_address`) and the original parameter (`user_street`) is not needed anymore.

**Comment by MarcMichalsky** [comment]
> This removes e.g. the user_street element from the $params array, since its value is being set to the internal CiviCRM parameter name (street_address) and the original parameter (user_street) is not needed anymore.

But the condition is always `true` because the array is hard coded.

**Comment by bjendres** [comment]
@jensschuppe, any objections to pulling this? It looks sensible and @mariav0 confirmed that it works for her... And afaik there's no other address parameters submitted by twingle.

**Comment by jensschuppe** [comment]
@bjendres I haven't merged it yet, because

* it changes current behavior (someone might rely on addresses containing country only)
* we should therefore  think about making this configurable per profile
* we should maybe flip the logic (if _nothing else than country_ instead of _not everything but country_ - with "everything" being hard-coded; i.e. use an `array_diff()` or similar)
* it does not check for empty address parameters (only for whether they exist with `array_key_exists()`

**Comment by MarcMichalsky** [comment]
@jensschuppe 
> it does not check for empty address parameters (only for whether they exist with array_key_exists()

As far as I know, Twingle does not submit empty fields. But I can check this, if you want.

**Comment by MarcMichalsky** [comment]
Okay, I checked it: **Twingle does submit empty address data fields**. So we should include another check

**Comment by jensschuppe** [comment]
@MarcMichalsky Could you rebase this on the current `master` since #53 changed some code in the `TwingleDonation.submit` API code and I'd like that to not interfere with your work.

**Comment by MarcMichalsky** [comment]
Okay, that should be it.
I have made the required address components configurable in the profile.

The check for empty address parameters is already done here:
https://github.com/systopia/de.systopia.twingle/blob/197563920adf31bbeabe324cc358401dc4034ff8/api/v3/TwingleDonation/Submit.php#L324-L336

File: api/v3/TwingleDonation/Submit.php [hunk 1]
```diff
@@ -347,7 +347,12 @@ function civicrm_api3_twingle_donation_Submit($params) {
         }
       }
       if ($unset_address_params) {
-        foreach($required_address_components as $req) {
```
Comment (MarcMichalsky) [inline api/v3/TwingleDonation/Submit.php]: Sorry, I had a knot in my brain here. 🧠

**Comment by jensschuppe** [comment]
I did some slight changes in 6779349505ba35a54fa5a69948c40d50b6d68dc3 and merged to `master`. This should be backwards-compatible due to the new configuration option. However, it defaults to requiring street, postal code, city and country for new profiles, but that seems reasonable to me.

@MarcMichalsky you might want to re-run your tests on that, but the code looks good to me. I went ahead merging it in order to  move on towards a pre-release.

**Pull request merged**